src/grisp_spi.erl

```diff
@@ -0,0 +1,192 @@
+% of the bytes being received. Other devices can start sending data already
+% before the controller has finished sending a request, and even reply with
+% data while the controller is writing its own data.
+%
+% This module provides two message formats. One format ({@link message_raw
+% ()}) returns the full response, all bytes received from the responder device
+% including ones sent during the initial request.
+%
+% The other format ({@link message_simple()}) automatically pads the request so
+% that response bytes can be received, and strips the response of an initial
+% number of bytes. This is makes it simple to use responder devices which wait
+% full for a received request before replying.
+-module(grisp_spi).
+
+-include("grisp_nif.hrl").
+
+% API
+-export([open/1]).
+-export([open/2]).
+-export([transfer/2]).
+
+% Callbacks
+-export([on_load/0]).
+-on_load(on_load/0).
+
+-define(CPOL_HIGH, 1).
+-define(CPHA_TRAILING, 2).
+-define(CS_DISABLE, 4).
+
+%--- Types ---------------------------------------------------------------------
+
+-type pin() :: {cs, integer()} | {gpio, grisp_gpio:bus()}.
+-type bus() :: spi1 | spi2.
+% SPI bus identifier.
+-type clock() :: {Polarity :: low | high, Phase :: leading | trailing}.
+% SPI clock configuration.
+-type mode() :: #{clock := clock()}.
+% SPI transfer mode.
+-type message_raw() :: {Mode::mode(), Message::binary()}.
+% A message where the request is unmodified and which will yield a response
+% binary of the same size.
+-type message_simple() :: {
+    Mode::mode(),
+    Message::binary(),
+    Skip::non_neg_integer(),
+    Pad::non_neg_integer()
+}.
+% A message where the request itself is padded with `Pad' number of `0' bytes,
+% and will yield a response binary that is stripped of its first `Skip'
+% number of bytes.
+-type message() :: message_raw() | message_simple().
+% SPI message.
+-type response() :: binary().
+% SPI response.
+-opaque ref() :: {reference(), pin()}.
+% Reference to an opened SPI bus.
+
+-export_type([bus/0]).
+-export_type([clock/0]).
+-export_type([mode/0]).
+-export_type([message/0]).
+-export_type([response/0]).
+-export_type([ref/0]).
+
+%--- API -----------------------------------------------------------------------
+
+% @doc Open an SPI bus with the default chip select pin.
+%
+% The respective pin 1 is used as `DefaultPin' for each slot.
+%
+% @equiv open(Slot, DefaultPin)
+% @see grisp_gpio
+-spec open(Bus::bus()) -> ref().
+open(spi1) -> open(spi1, spi1_pin1);
+open(spi2) -> open(spi2, spi2_pin1).
+
+% @doc Open an SPI bus.
+%
+% Chip select pin `CS' can be any valid GPIO output pin.
+%
+% @see grisp_gpio
+-spec open(bus(), grisp_gpio:pin()) -> ref().
+open(_Bus, CS) -> {open_nif(), pin(grisp_hw:platform(), CS)}.
+
+% @doc Transfer SPI messages on a bus.
+%
+% A message can either be a raw binary message ({@link message_raw()}), for
+% which a response of the same length is returned. Alternatively, it can be a
```

Comment:
The "either ... Alternatively,..." structure is messing up my reading flow. I am looking for the "or" from "either .. or".

Reply:
Please check the new text, I changed it around a bit.